### ISSUE TITLE
Bad PUT request, Failed to store item, 'NoneType' object has no attribute 'dtstart'

### ISSUE THREAD

**Issue by sbechet** [Kozea/Radicale#689]
Hello,

I'm new to radicale, currently using Baikal, but antipattern speak to me...

Radicale version is 2.1.4

Yesterday evening, i tried to import all my calendars from Evolution

and ...

```
 'wsgi.errors': <_io.StringIO object at 0x75a62f80>,
 'wsgi.file_wrapper': <class 'wsgiref.util.FileWrapper'>,
 'wsgi.input': <_io.BufferedReader name=6>,
 'wsgi.multiprocess': False,
 'wsgi.multithread': True,
 'wsgi.run_once': False,
 'wsgi.url_scheme': 'http',
 'wsgi.version': (1, 0)}
[75957470] DEBUG: Sanitized script name: ''
[75957470] DEBUG: Sanitized path: '/test/2a55c00a-2bb7-47ca-997b-f775a0dee85b/da907f28-b74f-4beb-bb71-67eaa70f621c-20170823T075351Z.ics'
[75957470] INFO: Successful login: 'test'
[75957470] DEBUG: Request content:
BEGIN:VCALENDAR
CALSCALE:GREGORIAN
PRODID:-//Ximian//NONSGML Evolution Calendar//EN
VERSION:2.0
BEGIN:VTIMEZONE
TZID:Europe/Paris
X-LIC-LOCATION:Europe/Paris
BEGIN:STANDARD
TZNAME:CET
DTSTART:19701029T030000
RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
END:STANDARD
BEGIN:DAYLIGHT
TZNAME:CEST
DTSTART:19700326T020000_RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VEVENT
DTSTAMP:20170626T190741Z
UID:da907f28-b74f-4beb-bb71-67eaa70f621c
RECURRENCE-ID:20170417T180000Z
SEQUENCE:1
DTSTART;TZID=Europe/Paris:20170417T200000
DTEND;TZID=Europe/Paris:20170417T213000
SUMMARY:This is a summary
STATUS:CANCELLED
CLASS:PUBLIC
LAST-MODIFIED:20170823T075353Z
END:VEVENT
END:VCALENDAR

[75957470] WARNING: Bad PUT request on '/test/2a55c00a-2bb7-47ca-997b-f775a0dee85b/da907f28-b74f-4beb-bb71-67eaa70f621c-20170823T075351Z.ics': Failed to store item 'da907f28-b74f-4beb-bb71-67eaa70f621c-20170823T075351Z.ics' in collection 'test/2a55c00a-2bb7-47ca-997b-f775a0dee85b': 'NoneType' object has no attribute 'dtstart'
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 1378, in upload
    href, vobject_item)
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 1207, in _store_item_cache
    content = self._item_cache_content(href, vobject_item, cache_hash)
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 1201, in _item_cache_content
    tag, start, end = xmlutils.find_tag_and_time_range(vobject_item)
  File "/usr/lib/python3.6/site-packages/radicale/xmlutils.py", line 635, in find_tag_and_time_range
    _visit_time_ranges(vobject_item, tag, range_fn, infinity_fn)
  File "/usr/lib/python3.6/site-packages/radicale/xmlutils.py", line 322, in _visit_time_ranges
    dtstart = child.dtstart.value
AttributeError: 'NoneType' object has no attribute 'dtstart'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 890, in do_PUT
    new_item = parent_item.upload(href, items[0])
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 1381, in upload
    (href, self.path, e)) from e
ValueError: Failed to store item 'da907f28-b74f-4beb-bb71-67eaa70f621c-20170823T075351Z.ics' in collection 'test/2a55c00a-2bb7-47ca-997b-f775a0dee85b': 'NoneType' object has no attribute 'dtstart'
[75957470] DEBUG: Response content:
Bad Request
[75957470] INFO: PUT response status for '/test/2a55c00a-2bb7-47ca-997b-f775a0dee85b/da907f28-b74f-4beb-bb71-67eaa70f621c-20170823T075351Z.ics' in 0.312 seconds: 400 Bad Request
```

Note: I already have an ics file with the same ID in the collection.

Ready to help. Ask me.

**Comment by sbechet** [comment]
```
test@mycomputer:~/.cache/evolution/calendar$ grep -ri da90 *
1430057603.6989.43@mycomputer/calendar.ics:UID:da907f28-b74f-4beb-bb71-67eaa70f621c
1430057603.6989.43@mycomputer/calendar.ics: /cal.php/calendars/test/default/da907f28-b74f-4beb-bb71-67eaa70f621c.ic
1430057603.6989.43@mycomputer/calendar.ics:UID:da907f28-b74f-4beb-bb71-67eaa70f621c
1430057603.6989.43@mycomputer/calendar.ics: /cal.php/calendars/test/default/da907f28-b74f-4beb-bb71-67eaa70f621c.ic
1430057603.6989.43@mycomputer/calendar.ics:UID:da907f28-b74f-4beb-bb71-67eaa70f621c
1430057603.6989.43@mycomputer/calendar.ics: /cal.php/calendars/test/default/da907f28-b74f-4beb-bb71-67eaa70f621c.ic
1503474327.1172.1@mycomputer/calendar.ics:UID:da907f28-b74f-4beb-bb71-67eaa70f621c
1503474327.1172.1@mycomputer/calendar.ics: /test/2a55c00a-2bb7-47ca-997b-f775a0dee85b/da907f28-b74f-4beb-bb71-67ea
```

with `1430057603.6989.43` = current baikal store and `1503474327.1172.1` = new radicale test.

**Comment by Unrud** [comment]
> Note: I already have an ics file with the same ID in the collection.

Your ICS files seem to be fragmented into multiple files. All components with the same UID must be in the same file.

> ``AttributeError: 'NoneType' object has no attribute 'dtstart'``

That error message is a bit unclear. It means that the main component was not found in the ICS file.
The ICS file only contains a overwritten recurrence.

> i tried to import all my calendars from Evolution

How did you do this?

**Comment by sbechet** [comment]
I understand. In your opinion what is the best solution. Should I manually refill the uid that causes me problems? Nobody ever has this problem?

I think I will try to extract from my calendar all the entries of the uid that are problematic to me and I will send it here so that we can exchange on this issue.

About gnome evolution, I configured the calendar to access Radicale and then I right clicked on my original calendar (which is otherwise synchronized to a Baikal server) and then copy all entries in the Radical calendar.

**Comment by sbechet** [comment]
Hello,

I take time to fully extract gnome evolution data from `~/.cache/evolution/calendar/1430057603.6989.43@mycomputer/calendar.ics` .

I found three VEVENT with the same GUID.

# First one

```
SUMMARY:A first action
SEQUENCE:3
DTSTART;TZID=Europe/Paris:20160905T200000
STATUS:CONFIRMED
```

When I create "A first action" for a long long time...
FREQ=WEEKLY;UNTIL=20170626T175959Z;BYDAY=MO

# Second one

```
SUMMARY:A first action
SEQUENCE:1
DTSTART;TZID=Europe/Paris:20170417T200000
STATUS:CANCELLED
```

I decide to CANCEL my recurrence.

# Last one

```
SUMMARY:Another summary ?
SEQUENCE:1
DTSTART;TZID=Europe/Paris:20170619T200000
STATUS:CONFIRMED
```

It seems a software take again GUID?

# Full VEVENT

```
BEGIN:VEVENT
DTSTAMP:20170626T190741Z
UID:da907f28-b74f-4beb-bb71-67eaa70f621c
SEQUENCE:3
DTSTART;TZID=Europe/Paris:20160905T200000
DURATION:PT1H30M
RRULE:FREQ=WEEKLY;UNTIL=20170626T175959Z;BYDAY=MO
SUMMARY:A first action
LOCATION:Somewhere
STATUS:CONFIRMED
CLASS:PUBLIC
CREATED:20160905T082937Z
X-MOZ-LASTACK:20170621T194726Z
X-MOZ-GENERATION:143
X-EVOLUTION-CALDAV-HREF:
 /cal.php/calendars/myusername/default/da907f28-b74f-4beb-bb71-67eaa70f621c.ic
 s
X-EVOLUTION-CALDAV-ETAG:"cd18876adb1bac4f96b56fd42130a2be"
BEGIN:VALARM
TRIGGER;VALUE=DURATION:-PT10M
ACTION:DISPLAY
DESCRIPTION:A first action
X-EVOLUTION-ALARM-UID:20170626T194117Z-1097-100-937-13@mycomputer
END:VALARM
END:VEVENT


BEGIN:VEVENT
DTSTAMP:20170626T190741Z
UID:da907f28-b74f-4beb-bb71-67eaa70f621c
RECURRENCE-ID:20170417T180000Z
SEQUENCE:1
DTSTART;TZID=Europe/Paris:20170417T200000
DTEND;TZID=Europe/Paris:20170417T213000
SUMMARY:A first action
STATUS:CANCELLED
CLASS:PUBLIC
X-EVOLUTION-CALDAV-HREF:
 /cal.php/calendars/myusername/default/da907f28-b74f-4beb-bb71-67eaa70f621c.ic
 s
X-EVOLUTION-CALDAV-ETAG:"cd18876adb1bac4f96b56fd42130a2be"
END:VEVENT


BEGIN:VEVENT
DTSTAMP:20170626T190741Z
UID:da907f28-b74f-4beb-bb71-67eaa70f621c
RECURRENCE-ID:20170619T180000Z
SEQUENCE:1
DTSTART;TZID=Europe/Paris:20170619T200000
DTEND;TZID=Europe/Paris:20170619T213000
SUMMARY:Another summary ?
STATUS:CONFIRMED
CLASS:PUBLIC
X-EVOLUTION-CALDAV-HREF:
 /cal.php/calendars/myusername/default/da907f28-b74f-4beb-bb71-67eaa70f621c.ic
 s
X-EVOLUTION-CALDAV-ETAG:"cd18876adb1bac4f96b56fd42130a2be"
BEGIN:VALARM
TRIGGER;VALUE=DURATION:-PT10M
ACTION:DISPLAY
DESCRIPTION:Another summary ?
X-EVOLUTION-ALARM-UID:20170626T194117Z-1097-100-937-14@mycomputer
END:VALARM
END:VEVENT
```

**Comment by Unrud** [comment]
> About gnome evolution, I configured the calendar to access Radicale and then I right clicked on my original calendar (which is otherwise synchronized to a Baikal server) and then copy all entries in the Radical calendar.

That feature of Evolution seems to be buggy. It doesn't copy all components with the same UID into the same calendar object. You might want to fill a bug.

> In your opinion what is the best solution.

In Evolution select ``Save as...`` in the right-click menu of your original calendar to export it as an ICS file. You can upload the calendar with **curl** to Radicale.
Example:
```shell
curl -u 'test' -X PUT 'http://localhost:5232/test/2a55c00a-2bb7-47ca-997b-f775a0dee85b/' --data-binary @'calendar.ics'
```
The existing calendar in Radicale gets overwritten.
If the calendar has a lot of events, it might fail with Radicale <= 2.1.4.

**Comment by sbechet** [comment]
Hello @Unrud,

Thank you for your help.

I copy calendar.ics (directly from `.cache/evolution/calendar/1430057603.6989.43\@mycomputer/calendar.ics`) then try your command

* First try, I had many time the line `ERROR:root:No module named 'pytz'`. I add python-pytz. Maybe it can be usefull to add it to Radicale dependencies?

* Second try, I had many time the line `ERROR:root:'/freeassociation.sourceforge.net/Europe/Paris'`, then i wait a looooong time (RPI3) then stack trace:

```
[75825470] ERROR: An exception occurred during PUT request on '/myuser/1923499e-ca56-2041-1090-fcb3e4222c58/': [Errno 24] Too many open files: '/home/radicale/.local/share/radicale/collection-root/myuser/.Radicale.tmp-oku1hoep/collection/.Radicale.cache/item/67bb0ce7-9f74-db6c-dc9d-a0c4000b6490.ics'
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 312, in __call__
    status, headers, answers = self._handle_request(environ)
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 483, in _handle_request
    environ, base_prefix, path, user)
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 877, in do_PUT
    path, items, props)
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 902, in create_collection
    self._upload_all_nonatomic(vobject_items)
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 951, in _upload_all_nonatomic
    open(os.path.join(cache_folder, href), "wb")))
OSError: [Errno 24] Too many open files: '/home/radicale/.local/share/radicale/collection-root/myuser/.Radicale.tmp-oku1hoep/collection/.Radicale.cache/item/67bb0ce7-9f74-db6c-dc9d-a0c4000b6490.ics'
ERROR:radicale:An exception occurred during PUT request on '/myuser/1923499e-ca56-2041-1090-fcb3e4222c58/': [Errno 24] Too many open files: '/home/radicale/.local/share/radicale/collection-root/myuser/.Radicale.tmp-oku1hoep/collection/.Radicale.cache/item/67bb0ce7-9f74-db6c-dc9d-a0c4000b6490.ics'
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 312, in __call__
    status, headers, answers = self._handle_request(environ)
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 483, in _handle_request
    environ, base_prefix, path, user)
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 877, in do_PUT
    path, items, props)
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 902, in create_collection
    self._upload_all_nonatomic(vobject_items)
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 951, in _upload_all_nonatomic
    open(os.path.join(cache_folder, href), "wb")))
OSError: [Errno 24] Too many open files: '/home/radicale/.local/share/radicale/collection-root/myuser/.Radicale.tmp-oku1hoep/collection/.Radicale.cache/item/67bb0ce7-9f74-db6c-dc9d-a0c4000b6490.ics'
```

It seems somewhere a close(handle) is missing.

Soooo, I kill radicale and run again with `radicale --debug` and try to import evolution file exactly like you ask me.

Here is the diff:

```
--- calendar.ics	2017-08-24 23:10:43.338370590 +0200
+++ test-record.ics	2017-08-24 23:34:13.078900510 +0200
@@ -3,21 +3,6 @@
 PRODID:-//Ximian//NONSGML Evolution Calendar//EN
 VERSION:2.0
 BEGIN:VTIMEZONE
-TZID:Romance Standard Time
-BEGIN:STANDARD
-DTSTART:16010101T030000
-TZOFFSETFROM:+0200
-TZOFFSETTO:+0100
-RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10
-END:STANDARD
-BEGIN:DAYLIGHT
-DTSTART:16010101T020000
-TZOFFSETFROM:+0100
-TZOFFSETTO:+0200
-RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3
-END:DAYLIGHT
-END:VTIMEZONE
-BEGIN:VTIMEZONE
 TZID:Europe/London
 X-LIC-LOCATION:Europe/London
 BEGIN:STANDARD
@@ -35,6 +20,42 @@
 TZOFFSETTO:+0100
 END:DAYLIGHT
 END:VTIMEZONE
+BEGIN:VTIMEZONE
+TZID:Europe/Paris
+X-LIC-LOCATION:Europe/Paris
+BEGIN:STANDARD
+TZNAME:CET
+DTSTART:19701029T030000
+RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10
+TZOFFSETFROM:+0200
+TZOFFSETTO:+0100
+END:STANDARD
+BEGIN:DAYLIGHT
+TZNAME:CEST
+DTSTART:19700326T020000
+RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3
+TZOFFSETFROM:+0100
+TZOFFSETTO:+0200
+END:DAYLIGHT
+END:VTIMEZONE
+BEGIN:VTIMEZONE
+TZID:/freeassociation.sourceforge.net/Europe/Paris
+X-LIC-LOCATION:Europe/Paris
+BEGIN:STANDARD
+TZNAME:CET
+DTSTART:19701029T030000
+RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10
+TZOFFSETFROM:+0200
+TZOFFSETTO:+0100
+END:STANDARD
+BEGIN:DAYLIGHT
+TZNAME:CEST
+DTSTART:19700326T020000
+RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3
+TZOFFSETFROM:+0100
+TZOFFSETTO:+0200
+END:DAYLIGHT
+END:VTIMEZONE
 BEGIN:VEVENT
 CREATED:20160729T092235Z
 LAST-MODIFIED:20160802T090701Z
```

This time:
```
[75854470] ERROR: An exception occurred during PUT request on '/myuser/1923499e-ca56-2041-1090-fcb3e4222c58/': [Errno 24] Too many open files: '/home/radicale/.local/share/radicale/collection-root/myuser/.Radicale.tmp-xim4rsr1/collection/.Radicale.cache/item/123f4c21-7603-5c02-e7b1-9450660518c5.ics'
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 312, in __call__
    status, headers, answers = self._handle_request(environ)
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 483, in _handle_request
    environ, base_prefix, path, user)
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 877, in do_PUT
    path, items, props)
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 902, in create_collection
    self._upload_all_nonatomic(vobject_items)
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 951, in _upload_all_nonatomic
    open(os.path.join(cache_folder, href), "wb")))
OSError: [Errno 24] Too many open files: '/home/radicale/.local/share/radicale/collection-root/myuser/.Radicale.tmp-xim4rsr1/collection/.Radicale.cache/item/123f4c21-7603-5c02-e7b1-9450660518c5.ics'
```

It is true that my calendar was started a long time ago and has been used by many different applications. In fact, I think these bugs should never happen in real life.

**Comment by Unrud** [comment]
> It seems somewhere a close(handle) is missing.

It's already fixed: 1821b872d2d144257fe94d7b8a449236dc3ef5bd

> If the calendar has a lot of events, it might fail with Radicale <= 2.1.4.

Version 2.1.5 will be released soon.

> First try, I had many time the line ``ERROR:root:No module named 'pytz'``. I add python-pytz. Maybe it can be usefull to add it to Radicale dependencies?

Radicale doesn't use ``pytz`` directly. The problem you encountered is probably in [vobject](https://github.com/eventable/vobject). You have to take a look at the stack trace.

**Comment by sbechet** [comment]
Hello @Unrud!

Good news, all is working perfectly with [1821b8](https://github.com/Kozea/Radicale/commit/1821b872d2d144257fe94d7b8a449236dc3ef5bd)
```
[7585f470] INFO: PUT response status for '/myuser/1923499e-ca56-2041-1090-fcb3e4222c58/' in 1454.546 seconds: 201 Created
```

Thank you _very_ much for your help and magic massive import `curl` command!
